design/pool-configuration.md

```diff
@@ -0,0 +1,359 @@
+* [Issue: allow selecting services to announce with the bgb speaker](https://github.com/metallb/metallb/issues/604)
+* [Issue: Dynamic BGP communities from services](https://github.com/metallb/metallb/issues/934)
+
+### Risks and Mitigations
+
+#### Conflicts with the current implementation
+
+The only current configuration option that will collide with what is being described here is the BGP peer node selector.
+
+Despite some overlap, it seems reasonable to maintain both filters and the only risk will be that if there is no intersection between filters,
+the service won't be advertised. No risk of non-deterministic behavior seems to be present.
+
+#### Bad configuration
+
+##### Service allocation
+
+When a user specify both a pool to be allocated to a specific service and one with no selector, the service may receive an IP from the broader one.
+We need to implement an easy to understand and deterministic behaviour (see design details below).
+
+##### Non overlapping filters
+
+Having an IP with no nodes advertising it is a risk. Some scenarios are:
+
+* Traffic Policy Local with pods not running on the same nodes selected by the selector
+* Node BGP peer selector with no intersection with the address selector
+
+It should be possible to detect non overlapping configurations, and raise an event when they happen. The metallb operator introduces a Validation WebHook
+that can be used to prevent those configurations to be created in first instance.
+
+### Conflicts with discovery
+
+The peer discovery feature currently being implemented in [this PR](https://github.com/metallb/metallb/pull/593) will introduce a way to
+fetch peers from the node's annotations.
+
+In order to be able to select them, the current proposed structure
+
+```yaml
+from-annotations:
+      - my-asn: example.com/local-asn
+        peer-asn: example.com/remote-asn
+        peer-address: example.com/address
+```
+
+will need to be extended with a name field (as much as the `peers` substructure) so it can be selected.
+
+### Conflicts with the FRR implementation
+
+All the BGP changes will happen from a "behavioural" point of view. Where needed, the changes will reflected in both the native and the new implementation.
+
+### Not covering all the angles
+
+The risk of not covering all possible cases / permutations is still there. For example, the proposed solution leverages pool customization to provide a
+way to set per service specific properties (i.e. define an AddressPool with bgp-communities and associate it to a single service), while a user may want to
+set custom properties only on one single service.
+
+That aside, this proposal covers the majority of the issues that were raised by the users (multiple times), and it won't prevent to add a finer level of configuration
+if ever comes the need.
+
+## Design Details
+
+### Announcing the IP from a subset of nodes
+
+Extending the address-pool object with an optional node selector field is straightforward:
```

Comment:
Consider the following use case:

- **IP X** is allocated to some service.
- **Node A** is peered with **routers 1 and 2**. **Node B** exists as well.
- I want to advertise IP X **from node A** (and not node B) but **only to router 1**.

I'm not sure we can handle this use case in the proposed design, but even if we can, I'd say the following design would be more "natural" when thinking about how this kind of policy is handled by commercial routers:

We apply a policy **to node A**. This policy filters BGP advertisements to routers 1 and 2, allowing IP X to be advertised to router 1 but not to router 2.

"Translating" the above to MetalLB terms, I'd consider putting the "filtering" logic in the **peer**:

```yaml
    peers:
    - peer-address: 10.0.0.1
      peer-asn: 64501
      my-asn: 64500
      name: foo
      allowed-prefixes:
      - 172.16.0.1/32
      - 192.168.1.0/29
```

In case you want to have two different policies for node A and node B *for the same peer 10.0.0.1*, you would define **two peer sections**, each with a different node selector, with the same `peer-address` and different `allowed-prefixes` clauses.

WDYT?

Reply:
The scenario you are describing is covered with the next section, "Announcing the IP to a subset of BGP neighbors".

The first thought that comes into my mind is that this won't work with L2, and a lot of users asked for this feature because in L2 only one node advertises and different nodes may be connected to different networks.

Said that, I am a bit torn about this. On one hand, it's more flexible and near to what it will look like in FRR.

On the other hand, it feels a bit too low level. Users will have to configure the addresspools, and then go manually and copy the cidrs to the peers. It feels more natural to say "I want this peer to advertise this pool only" (or as per my version, "I want this pool to be advertised only to these peers") because the pool is the higher level object the prefixes come from.

I saw value in having the addresspool the central entity of our configuration. Together with service mapping logic, it would allow metallb to be used in multi tenant scenarios that require different configurations for different tenants. 

With this, it would be doable, but more fragmented and error prone because we'd loose the natural association with the pool and with the namespace.

Would that work if `allowed-prefixes` accepts also strings and in that case the semantic is that it's the name of a pool? Or having an additional `allowed-pools` field together with that (maybe too many knobs)? We'd loose the addressPool being the center of the configuration, but we'd still be able to provide a higher level abstraction that will reduce configuration errors.

Reply:
You've raised good points.

Following the discussion in the community meeting and since we're considering implementing this design using CRDs anyway, I think the best approach may be to separate the concerns of **allocating** IPs and **advertising** IPs. If we take this approach, the following happens:

- An address pool object becomes protocol-agnostic and specifies only allocation-related params
- Either we create a layer 2 equivalent of `peer` and control prefix filtering there (by referencing address pools), **or** we add a protocol-agnostic abstraction, say `kind: Advertisement`, which controls whether or not to advertise a prefix to a given router (and then we don't need a `peer` equivalent for layer 2 as long as there are no APR-specific knobs to tweak).

Reply:
Yep, I'll rework this proposal to include a re-design of the data structures (I feel like it won't change to much whether we implement them as sections of the configmap or as separate crds), so we can iterate from there (next year :P ).